cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputCommitter.java

```diff
@@ -0,0 +1,237 @@
+    Partitioning partitioning = outputDataset.getPartitioning();
+
+    Set<PartitionKey> partitionsToAdd = new HashSet<>();
+    // Go over all files in the temporary directory and keep track of partitions to add for them
+    FileStatus[] allCommittedTaskPaths = getAllCommittedTaskPaths(context);
+    for (FileStatus committedTaskPath : allCommittedTaskPaths) {
+      FileSystem fs = committedTaskPath.getPath().getFileSystem(context.getConfiguration());
+      RemoteIterator<LocatedFileStatus> fileIter = fs.listFiles(committedTaskPath.getPath(), true);
+      while (fileIter.hasNext()) {
+        Path path = fileIter.next().getPath();
+        String relativePath = getRelative(committedTaskPath.getPath(), path);
+
+        int lastPathSepIdx = relativePath.lastIndexOf(Path.SEPARATOR);
+        if (lastPathSepIdx == -1) {
+          // this shouldn't happen because each relative path should consist of at least one partition key and
+          // the output file name
+          LOG.warn("Skipping path '{}'. It's relative path '{}' has fewer than two parts", path, relativePath);
+          continue;
+        }
+        // relativePath = "../key1/key2/part-m-00000"
+        // relativeDir = "../key1/key2"
+        // fileName = "part-m-00000"
+        String relativeDir = relativePath.substring(0, lastPathSepIdx);
+        String fileName = relativePath.substring(lastPathSepIdx + 1);
+
+        Path finalDir = new Path(FileOutputFormat.getOutputPath(context), relativeDir);
+        Path finalPath = new Path(finalDir, fileName);
+        if (fs.exists(finalPath)) {
+          throw new FileAlreadyExistsException("Final output path " + finalPath + " already exists");
+        }
+        PartitionKey partitionKey = getPartitionKey(partitioning, relativeDir);
+        partitionsToAdd.add(partitionKey);
+      }
+    }
+
+    // We need to copy to the parent of the FileOutputFormat's outputDir, since we added a _temporary_jobId suffix to
+    // the original outputDir.
+    Path finalOutput = FileOutputFormat.getOutputPath(context);
+    FileSystem fs = finalOutput.getFileSystem(context.getConfiguration());
+    for (FileStatus stat: getAllCommittedTaskPaths(context)) {
+      mergePaths(fs, stat, finalOutput);
+    }
+
+    // create all the necessary partitions
+    for (PartitionKey partitionKey : partitionsToAdd) {
+      outputDataset.getPartitionOutput(partitionKey).addPartition();
+    }
+    // close the TaskContext, which flushes dataset operations
+    try {
+      mrTaskContext.flushOperations();
+    } catch (Exception e) {
+      throw new IOException(e);
+    } finally {
+      mrTaskContext.close();
+    }
+
+    // delete the job-specific _temporary folder and create a _done file in the o/p folder
+    cleanupJob(context);
+    // True if the job requires output.dir marked on successful job.
+    // Note that by default it is set to true.
+    if (context.getConfiguration().getBoolean(SUCCESSFUL_JOB_OUTPUT_DIR_MARKER, true)) {
```

Comment:
why? For partitions that are written without dynamic partitioning, the _SUCCESS file will be there. Should dynamic partitions be different?